extensions/amp-ad/0.1/amp-ad-api-handler.js

```diff
@@ -148,29 +151,39 @@ export class AmpAdApiHandler {
 
   /**
    * Updates the element's dimensions to accommodate the iframe's
-   *    requested dimensions.
+   * requested dimensions. Notifies the window that request the resize
+   * of success or failure.
    * @param {number|undefined} height
    * @param {number|undefined} width
+   * @param {!Window} source
+   * @param {string} origin
    * @private
    */
-  updateSize_(height, width) {
-    const targetOrigin =
-          this.iframe_.src ? parseUrl(this.iframe_.src).origin : '*';
-    this.baseInstance_.attemptChangeSize(height, width, () => {
-      postMessage(
-          this.iframe_,
-          'embed-size-changed',
-          {requestedHeight: height, requestedWidth: width},
-          targetOrigin,
-          this.is3p_);
-    }, () => {
-      postMessage(
-          this.iframe_,
-          'embed-size-denied',
-          {requestedHeight: height, requestedWidth: width},
-          targetOrigin,
-          this.is3p_);
-    });
+  updateSize_(height, width, source, origin) {
+    this.baseInstance_.attemptChangeSize(height, width,
+        () => this.sendEmbedSizeResponse_(
+            true /* success */, width, height, source, origin),
+        () => this.sendEmbedSizeResponse_(
+            false /* success */, width, height, source, origin));
+  }
+
+  /**
+   * Sends a response to the window which requested a resize.
+   * @param {boolean} success
+   * @param {number} requestedWidth
+   * @param {number} requestedHeight
+   * @param {!Window} source
+   * @param {string} origin
+   * @private
+   */
+  sendEmbedSizeResponse_(
+      success, requestedWidth, requestedHeight, source, origin) {
+    postMessageToWindows(
+        this.iframe_,
+        [{win: source, origin}],
```

Comment:
So, I'm joining this a bit late and I see (most likely) the original GitHub issue in #3019. Could you please catch me up:

1. Is #3019 indeed the origin issue?
2. Why did we decide to allow `amp-ad` to communicate with child frames directly vs proxying these messages with 3p iframe (most immediate child of `amp-ad`)? I'm not saying that this is wrong, I'd just like to get into right context on this.